linkerd/app/core/src/errors.rs

```diff
@@ -35,20 +36,81 @@ pub enum Reason {
     Unexpected,
 }
 
-#[derive(Debug)]
-pub struct NewRespond<B>(std::marker::PhantomData<fn() -> B>);
+#[derive(Copy, Clone, Debug)]
+pub struct NewRespond(());
 
 #[derive(Copy, Clone, Debug)]
-pub enum Respond<B> {
-    Http1(http::Version, std::marker::PhantomData<fn() -> B>),
+pub enum Respond {
+    Http1(http::Version),
     Http2 { is_grpc: bool },
 }
 
-impl<A, B: Default> respond::NewRespond<http::Request<A>> for NewRespond<B> {
-    type Response = http::Response<B>;
-    type Respond = Respond<B>;
+pub enum ResponseBody<B> {
+    NonGrpc(B),
+    Grpc {
+        inner: B,
+        trailers: Option<http::HeaderMap>,
+    },
+}
+
+impl<B: hyper::body::Payload> hyper::body::Payload for ResponseBody<B>
+where
+    B::Error: Into<Error>,
+{
+    type Data = B::Data;
+    type Error = B::Error;
+
+    fn poll_data(&mut self) -> Poll<Option<Self::Data>, Self::Error> {
+        match self {
+            Self::NonGrpc(inner) => inner.poll_data(),
+            Self::Grpc { inner, trailers } => {
+                // should not be calling poll_data if we have set trailers derived from an error
+                assert!(trailers.is_none());
+                match inner.poll_data() {
+                    Err(error) => {
+                        let error = error.into();
+                        let mut error_trailers = http::HeaderMap::new();
+                        let code = set_grpc_status(&error, &mut error_trailers);
+                        debug!(%error, grpc.status = ?code, "Handling gRPC stream failure");
+                        *trailers = Some(error_trailers);
+                        Ok(Async::Ready(None))
+                    }
+                    data => data,
+                }
+            }
+        }
+    }
+
+    fn poll_trailers(&mut self) -> futures::Poll<Option<http::HeaderMap>, Self::Error> {
+        match self {
+            Self::NonGrpc(inner) => inner.poll_trailers(),
+            Self::Grpc { inner, trailers } => match trailers.take() {
+                Some(t) => Ok(Async::Ready(Some(t))),
+                None => inner.poll_trailers(),
+            },
+        }
+    }
+    fn is_end_stream(&self) -> bool {
```

Comment:
there should be a newline here. i'm surprised rustfmt doesn't complain..

Reply:
yes me as well actually.